modules/raft-api/src/main/java/org/apache/ignite/internal/raft/Command.java

```diff
@@ -17,10 +17,27 @@
 
 package org.apache.ignite.internal.raft;
 
+import org.apache.ignite.internal.hlc.HybridTimestamp;
 import org.apache.ignite.internal.network.NetworkMessage;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * A marker interface for replication group command.
  */
 public interface Command extends NetworkMessage {
+    /**
+     * This is called before a command is submitted to replication pipeline.
+     *
+     * @param safeTs Safe timestamp.
+     */
+    default void patch(HybridTimestamp safeTs) {}
+
+    /**
+     * Holds request's initiator timestamp.
+     *
+     * @return The timestamp.
+     */
+    default @Nullable HybridTimestamp initiatorTime() {
```

Comment:
1. Do you mean that ReadCommand should also be initiatorTime aware?
2. Do you mean that all WriteCommand should be initiatorTime aware?
3. Why it's @Nullable?

Reply:
1. optional
2. optional
3. this is nullable because this field is optional (protocol dependent). but it will always be not null for raft + writecommand.

Reply:
1. No. I've moved initiatorTime to WriteCommand.
2. It's valid. It can be null for CMG for example and any other group not requring mvcc extensions.
So it's not possible to remove @nullable.
Correctness is ensured by SafeTimestampPropagatingCommand, where this annotation is removed, so it's not possible to create SafeTimestampPropagatingCommand without initialTime.

Reply:
As far as I understand SafeTimePropagatingCommand should always have initiatorTime and any other commands should never have initiatorTime, is that correct? If true, adding @Nullable initiatorTime to WriteCommand seems like a bad design to me. I do understand that it's likely because of patching code in NodeImpl
```
            // Patch the command.
            if (event.done instanceof WriteCommandClosure) {
                WriteCommandClosure clo = (WriteCommandClosure) event.done;
                WriteCommand command = clo.command();
                HybridTimestamp timestamp = command.initiatorTime();
```
I guess that since you move safe time logic into raft you should also move `SafeTimePropagatingCommand` from
`package org.apache.ignite.internal.replicator.command;`
to 
`package org.apache.ignite.internal.raft;`

Reply:
I guess yes. Since ST logic is now part of raft it seems reasonable to move SafeTimeCommand there.

Reply:
Created a ticket
https://issues.apache.org/jira/browse/IGNITE-24143